File: solutions/java/leap/4/src/main/java/Leap.java
@FunctionalInterface
interface LeapYearChecker { 
    boolean isLeapYear(int year); 
}

class Leap {

    protected LeapYearChecker checker;

    public Leap() {
        checker = (year) -> year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
    }

    public boolean isLeapYear(int year) {
        return checker.isLeapYear(year);
    }
    
}

class GregorianLeap extends Leap {}

class JulianLeap extends Leap {

    public JulianLeap() {
        checker = (year) -> year % 4 == 0;
    }
    
}

File: solutions/julia/leap/1/leap.jl
"""
    is_leap_year(year)

Return `true` if `year` is a leap year in the gregorian calendar.

"""
function is_leap_year(year)
    modulo_by(n) = year % n == 0
    modulo_by(4) && !modulo_by(100) || modulo_by(400)
end

